File: client-scalaz/src/main/scala/org/openapitools/client/api/ServiceType.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import org.joda.time.LocalTime

import ServiceType._

case class ServiceType (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
business: Option[Retailer],
hub: Option[ServiceHub],
name: Option[String],
additionalLoadTime: Option[Long],
pickupAddressOverride: Option[Address],
dropoffAddressOverride: Option[Address],
pickupStartTimeOverride: Option[LocalTime],
pickupEndTimeOverride: Option[LocalTime],
dropoffStartTimeOverride: Option[LocalTime],
dropoffEndTimeOverride: Option[LocalTime],
dateOverride: Option[DateOverride])

object ServiceType {
  import DateTimeCodecs._
  sealed trait DateOverride
  case object NONE extends DateOverride
  case object PICKUPSTART extends DateOverride
  case object PICKUPEND extends DateOverride
  case object DROPOFFSTART extends DateOverride
  case object DROPOFFEND extends DateOverride

  object DateOverride {
    def toDateOverride(s: String): Option[DateOverride] = s match {
      case "NONE" => Some(NONE)
      case "PICKUPSTART" => Some(PICKUPSTART)
      case "PICKUPEND" => Some(PICKUPEND)
      case "DROPOFFSTART" => Some(DROPOFFSTART)
      case "DROPOFFEND" => Some(DROPOFFEND)
      case _ => None
    }

    def fromDateOverride(x: DateOverride): String = x match {
      case NONE => "NONE"
      case PICKUPSTART => "PICKUPSTART"
      case PICKUPEND => "PICKUPEND"
      case DROPOFFSTART => "DROPOFFSTART"
      case DROPOFFEND => "DROPOFFEND"
    }
  }

  implicit val DateOverrideEnumEncoder: EncodeJson[DateOverride] =
    EncodeJson[DateOverride](is => StringEncodeJson(DateOverride.fromDateOverride(is)))

  implicit val DateOverrideEnumDecoder: DecodeJson[DateOverride] =
    DecodeJson.optionDecoder[DateOverride](n => n.string.flatMap(jStr => DateOverride.toDateOverride(jStr)), "DateOverride failed to de-serialize")

  implicit val ServiceTypeCodecJson: CodecJson[ServiceType] = CodecJson.derive[ServiceType]
  implicit val ServiceTypeDecoder: EntityDecoder[ServiceType] = jsonOf[ServiceType]
  implicit val ServiceTypeEncoder: EntityEncoder[ServiceType] = jsonEncoderOf[ServiceType]
}

File: client-clojure/src/sirqul_io_t_platform/specs/service_type.clj
(ns sirqul-io-t-platform.specs.service-type
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.retailer :refer :all]
            [sirqul-io-t-platform.specs.service-hub :refer :all]
            [sirqul-io-t-platform.specs.address :refer :all]
            [sirqul-io-t-platform.specs.address :refer :all]
            [sirqul-io-t-platform.specs.local-time :refer :all]
            [sirqul-io-t-platform.specs.local-time :refer :all]
            [sirqul-io-t-platform.specs.local-time :refer :all]
            [sirqul-io-t-platform.specs.local-time :refer :all]
            )
  (:import (java.io File)))


(def service-type-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :business) retailer-spec
   (ds/opt :hub) service-hub-spec
   (ds/opt :name) string?
   (ds/opt :additionalLoadTime) int?
   (ds/opt :pickupAddressOverride) address-spec
   (ds/opt :dropoffAddressOverride) address-spec
   (ds/opt :pickupStartTimeOverride) local-time-spec
   (ds/opt :pickupEndTimeOverride) local-time-spec
   (ds/opt :dropoffStartTimeOverride) local-time-spec
   (ds/opt :dropoffEndTimeOverride) local-time-spec
   (ds/opt :dateOverride) string?
   })

(def service-type-spec
  (ds/spec
    {:name ::service-type
     :spec service-type-data}))
